pkg/csi/cinder/openstack/openstack_volumes.go

```diff
@@ -83,22 +84,37 @@ func (os *OpenStack) CreateVolume(name string, size int, vtype, availability str
 
 // ListVolumes list all the volumes
 func (os *OpenStack) ListVolumes(limit int, startingToken string) ([]volumes.Volume, string, error) {
-	nextPageToken := ""
+	var nextPageToken string
+	var vols []volumes.Volume
+
 	opts := volumes.ListOpts{Limit: limit, Marker: startingToken}
-	pages, err := volumes.List(os.blockstorage, opts).AllPages()
-	if err != nil {
-		return nil, nextPageToken, err
-	}
-	vols, err := volumes.ExtractVolumes(pages)
-	if err != nil {
-		return nil, nextPageToken, err
-	}
-	nextPageURL, err := pages.NextPageURL()
-	if err != nil && nextPageURL != "" {
-		if queryParams, nerr := url.ParseQuery(nextPageURL); nerr != nil {
+	err := volumes.List(os.blockstorage, opts).EachPage(func(page pagination.Page) (bool, error) {
+		var err error
+
+		vols, err = volumes.ExtractVolumes(page)
+		if err != nil {
+			return false, err
+		}
+
+		nextPageURL, err := page.NextPageURL()
+		if err != nil {
+			return false, err
+		}
+
+		if nextPageURL != "" {
+			queryParams, err := url.ParseQuery(nextPageURL)
+			if err != nil {
+				return false, err
+			}
 			nextPageToken = queryParams.Get("marker")
 		}
+
+		return false, nil
```

Comment:
@kayrus trying to understand here, dont we need to return true here? false will come out of List Block?

Reply:
correct. if we return true, then the ForEach loop will be continued

Reply:
https://pkg.go.dev/github.com/gophercloud/gophercloud/pagination#Pager.EachPage
```
EachPage iterates over each page returned by a Pager, yielding one at a time to a handler function. Return "false" from the handler to prematurely stop iterating.
```

Reply:
@kayrus Thanks. understood. you might have tested , but just to confirm

1. From the attacher , MaxEntries value can be 0, if not set. So when limit is 0, it by default returns all the items in the list right?
2. If NextPageURL returns error, then also it should be ok to return volumes.
    Like instead of L114-116 , we can directly return L118
          return vols, nextPageToken, err

Reply:
You're right, by default the limit is 0, in this case openstack API will return all items without a pagination.
If `NextPageURL` returns an error (malformed json), then an error should be returned. The same error will be returned even earlier, when `ExtractSnapshots` is called.

Reply:
right, if ExtractVolumes() return error , anyway vols value would be nil/empty. So that statement should suffice. Either way should be fine I suppose.